Add `recurse` method to `TailCalls`

tmccombs opened the pull request:
This method is a simple convenience that allows you to recursively call an anonymous function, by passing a wrapper around that function as an argument. This can be used as a looping construct.

tmccombs committed:
Add `recurse` method to `TailCalls`

This method is a simple convenience that allows you to recursively
call an anonymous function, by passing a wrapper around that function as
an argument. This can be used as a looping construct.

tmccombs commented on src/library/scala/util/control/TailCalls.scala:
```diff
@@ -111,4 +111,31 @@ object TailCalls {
+   * import scala.util.control.TailCalls._
+   *
+   * def fib(n: Int): Int =
+   *    recurse[(Int, Int, Int), Int]((n, 1, 0)) ( recur => { case (i, n1, n2) =>
```
It's unfortunate that you have to explicitly give the type parameters, but I wasn't able to figure out a way to get scala to infer the types (at least in scala 2, maybe 3 can do it?)

Needing an extra set of braces around the case is a little annoying too.

bishabosha replied:
You can make a Recursor class that captures the initial state and infers the type, then that class has an apply method accepts the lambda

tmccombs committed:
Change recures to use helper class

This allows slightly better type inference. Although in many cases you
still need to specify the return type (using somewhat unusual syntax).

SethTisue commented:
This looks plausible, but we cannot currently add new methods to the Scala 2 standard library, as per https://github.com/scala/scala-dev/issues/661

tmccombs commented:
FWIW, in scala 3 it might be worth implementing with a macro, so it can take an arbitrary number of arguments instead of exactly 1
